include/rmm/cuda_device.hpp

```diff
@@ -42,19 +42,58 @@ struct cuda_device_id {
   value_type id_;
 };
 
-namespace detail {
 /**
  * @brief Returns a `cuda_device_id` for the current device
  *
  * The current device is the device on which the calling thread executes device code.
  *
  * @return `cuda_device_id` for the current device
  */
-inline cuda_device_id current_device()
+inline cuda_device_id get_current_cuda_device()
 {
-  int dev_id{};
-  RMM_CUDA_TRY(cudaGetDevice(&dev_id));
+  cuda_device_id::value_type dev_id{};
+  RMM_ASSERT_CUDA_SUCCESS(cudaGetDevice(&dev_id));
   return cuda_device_id{dev_id};
 }
-}  // namespace detail
+
+/**
+ * @brief Returns the number of CUDA devices in the system
+ *
+ * @return Number of CUDA devices in the system
+ */
+inline int get_num_cuda_devices()
+{
+  cuda_device_id::value_type num_dev{};
+  RMM_ASSERT_CUDA_SUCCESS(cudaGetDeviceCount(&num_dev));
+  return num_dev;
+}
+
+/**
+ * @brief RAII class that sets the current CUDA device to the specified device on construction
+ * and restores the previous device on destruction.
+ */
+struct cuda_set_device_raii {
```

Comment:
**note (non-blocking):** In CUB's version of the equivalent type, it looks like it attempts to avoid an extraneous `cudaSetDevice` call in the dtor if the new device was the same as the old: https://github.com/NVIDIA/cccl/blob/f2b931bfddd1d06bd96ca9d65506625144256f78/cub/cub/util_device.cuh#L92-L110

No idea if this was just premature optimization or actually useful.

Reply:
Fixed, thanks.